Rudimentary docker network support

dcbw opened the pull request:
Start adding support for docker's network API (which is just a passthrough to libnetwork's API).

peter-edge commented:
Hey, just for reference, can you link the github.com/docker/docker corresponding code/docs? If you don't have time, I can check it out soon.

peter-edge commented on network.go:
```diff
@@ -0,0 +1,76 @@
+
+// ListNetworks returns all networks.
+func (c *Client) ListNetworks() ([]Network, error) {
+	path := "/networks"
```
nit: might as well just inline this :)

dcbw replied:
Sure.  I removed Service too, since it's currently unused.

peter-edge commented on network.go:
```diff
@@ -0,0 +1,76 @@
+
+// NetworkInfo returns information about a network by its ID.
+func (c *Client) NetworkInfo(id string) (*Network, error) {
+	path := "/networks/" + id
```
same comment

peter-edge commented on network.go:
```diff
@@ -0,0 +1,76 @@
+		return nil, err
+	}
+	var networks []Network
+	err = json.Unmarshal(body, &networks)
```
scope this (and below):

```
if err := json.Unmarshal(body, &networks); err != nil {
  return nil, err
}
```

peter-edge commented on network.go:
```diff
@@ -0,0 +1,76 @@
+// NoSuchNetwork is the error returned when a given network does not exist.
+type NoSuchNetwork struct {
+	ID  string
+	Err error
```
do you ever use Err? it doesn't look like it here.

dcbw replied:
I mainly copied NoSuchContainer, which also has Err and doesn't appear to use it.  I'm happy to take it out of NoSuchNetwork, but should I be consistent with NoSuchContainer, or should I remove the unused bit?

peter-edge replied:
Remove both, thanks :)

peter-edge commented on network.go:
```diff
@@ -0,0 +1,76 @@
+	if err.Err != nil {
+		return err.Err.Error()
+	}
+	return "No such network: " + err.ID
```
nit: i usually like `fmt.Sprintf("foo: %s", err.ID)` but no biggie

peter-edge commented:
Also overall: needs some testing :)

peter-edge commented on container.go:
```diff
@@ -135,6 +135,8 @@ type NetworkSettings struct {
 	Bridge      string                 `json:"Bridge,omitempty" yaml:"Bridge,omitempty"`
 	PortMapping map[string]PortMapping `json:"PortMapping,omitempty" yaml:"PortMapping,omitempty"`
 	Ports       map[Port][]PortBinding `json:"Ports,omitempty" yaml:"Ports,omitempty"`
+	NetworkID   string                 `json:"NetworkID,omitempty" yaml:"NetworkID,omitempty"`
```
just making sure - is it `json:"NetworkID"` or `json:"NetworkId"` for docker?

dcbw replied:
It's "NetworkID":

        "MacAddress": "02:42:ac:11:2a:02",
        "NetworkID": "b6408b405133c5475bfde2ad383b80732326e1fb9bc713f6586552836ba2fdd5",
        "PortMapping": null,
        "Ports": {},
        "SandboxKey": "/var/run/docker/netns/23659bc4a779",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null

peter-edge replied:
SGTM

dcbw commented:
Repushed with some fixups.

peter-edge commented:
Thanks! Still needs to address the string inlining, linking to docker/docker code, and tests :)

dcbw commented:
FWIW the JSON looks like this:

dcbw commented:
FWIW the JSON for Network looks like this:

&{Name:purple ID:2585945e1bb6f36514dde1c563dd8d078bc058d9a4953a0fd8780b3b5237e983 Type:cni Endpoints:[]}

dcbw commented:
Repushed with inlined string and added links to docker code.  No documentation yet on the API that I can find (not surprising, it's really new).  Still have to do some tests; but does the code look OK so far now?

fsouza commented:
@peter-edge thanks for reviewing this!

@dcbw except for the lack of tests, I think the code looks good for now. Regarding the Docker documentation issue, are these APIs available in Docker 1.8? If so, they're probably going to add some docs soon.

dcbw commented:
@fsouza these APIs are available in docker 1.8 if you enable the EXPERIMENTAL flag when building docker.  It seems very likely they'll get turned on by default in the near future though.

peter-edge commented:
@fsouza i think it's good to merge but could use at least some base for testing so that when this code is inevitably expanded, other OSS contributors know where to add tests. plus some documentation about the need to do EXPERIMENTAL/1.8 etc. @dcbw are you willing to do this? :)

dcbw commented:
@peter-edge yes, I'm willing to work on the tests in the next few days.

fsouza commented:
Cool, we can wait for the tests then!

@dcbw thank you!

dcbw commented:
Ok, added some testcases, do they look complete enough?  Also added a note to the README about the network API and DOCKER_EXPERIMENTAL, and added some more fields to the NetworkSettings struct that are present in docker 1.8.

peter-edge commented:
Looks great! @fsouza

fsouza commented:
I'm missing the tests for the CreateNetwork method. @dcbw any special reason for not having tests for this method?

Thank you! And sorry for the late response.

dcbw commented:
@fsouza no, just an oversight, sorry!  Added the client-side CreateNetwork tests and repushed.

fsouza commented:
@dcbw thank you!
